astropy/coordinates/attributes.py

```diff
@@ -278,6 +278,9 @@ class QuantityAttribute(Attribute):
     """
 
     def __init__(self, default=None, secondary_attribute='', unit=None, shape=None):
+        if default is None:
```

Comment:
Why not just remove the `=None` in the initializer, and then for good measure do
```
self.unit = unit
self.shape = shape
default = self.convert_input(default)[0]
super().__init__(default, secondary_attribute)
```
to let `Quantity` and shape-checks give their relevant error message.

Reply:
p.s. We might want to do this more generally, allowing `value is None` as done in `TimeAttribute`. But that would be for another PR.

Reply:
My thinking here is that if we remove the `=None` in the initializer, the user gets the more generic ``... missing 1 required positional argument ...`` message, and by doing it this way a more specific error message gets generated. I think this is desirable for maybe a version or two in the (unlikely, but possible) event that someone actually is using this somehow to create a frame and will be surprised why they are suddenly getting an ambiguous error message.

Also the "second half" of #8276 would need to be implemented this way, seeing as how the Quantity initializer fails when you give it a `None`, so this seems like a more natural transition to that.

Reply:
Not sure - it seems just extra code for no real purpose (since clearly `default=None` currently breaks, and we do not want to encourage to relying on any default). But no big deal.

Independently, I would suggest to pass `default` through `self.convert_input` - that way any non-quantity default gets caught properly upon initialization.

Actually, how about moving the error check there? Then it is right where we later will want to change the meaning of `None` (as in `TimeAttribute`)

Reply:
Yea, my concern is that someone might check the init signature, see that the default is `default=None` (confusing!) and expect that to work, but then immediately gets an error. I think I'm +0.1 to @mhvk's suggestion of just removing the TypeError and turning the error you have into a comment.

Reply:
To try to get something in for the next version I'm going to take @mhvk's *second* suggestion here. But I'll make a PR now for 3.2 to change  `default=None` -> `default`, as I see @adrn's point but want to give at least a bit of a chance for people to see the "newer" warning.